### PR TITLE
Add support for examples in OpenAPI block

### PR DESCRIPTION
**Pull request by vibhanshub**
Same description as https://github.com/GitbookIO/gitbook/pull/2313

### CODE REVIEW

**Review by Eloutmanixx — APPROVED** [review]

**Review by Eloutmanixx — APPROVED** [review]
xxxxxxxxxx

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISchema.tsx:96]: `example` is not supposed to be markdown so why use markdown here? (https://swagger.io/docs/specification/adding-examples/)

It can result in problems, ex: if an `example` contains a markdown string, it'll be processed there when the user expects to see the raw string.

We want to render a value instead of markdown text.

Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: Also the spacing with the description is not great, we should use a gap between the 2

**Review by SamyPesse — CHANGES_REQUESTED** [review]

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: This is what it looks like now
Screen Shot 2024-05-15 at 2 20 31 PM

Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: Using
Screen Shot 2024-05-15 at 2 28 24 PM
 margin instead of padding now

Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: Also, argos is complaining for the right reasons as Examples are not present in argos. Can I approve it or do I need to update the images argos is using as a baseline? (first time working with argos where errors are justified)

Comment (scazan) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: If the change is expected and the new version you see in Argos is the correct one, simply approve it and it'll update the baseline.

Comment (taranvohra) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: The font styling looks off to me. Have you checked that?

**Review by scazan — APPROVED** [review]

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (taranvohra) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: The text color seems to stand about a bit too much. Is it black (#000)? Is it the same text color as the property name (`iso-id` in your screenshot)

Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: This is what it looks like with the same color as property name

```
.openapi-schema-example {
    @apply mt-2 text-base text-dark/10 dark:text-light/10;
}
```

Screen Shot 2024-05-20 at 4 50 00 PM

Looks the same to me

Comment (taranvohra) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: Idk why but this looks a bit better to me haha. It's subtle but it doesn't come across as fully black. Anyway, lgtm otherwise

**Review by taranvohra — APPROVED** [review]

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISchema.tsx:17]: I don't understand why we need this prop. Isn't `example` included in the `schema` prop (https://github.com/kogosoftwarellc/open-api/blob/main/packages/openapi-types/index.ts#L166) ?

Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:17]: Ok it looks like `example` is for parameters and `examples` is for schema fields. I think we should try to handle both the same way.

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISchema.tsx:97]: We should use a `<code>` or `<pre>` around the value.

Also, how does it work if my example is not a primitive?

ex:

```
example:
  propA: a
  propB: b
```

My take is that it'll render `[object Object]` which is bad.

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:17]: It looks like you don't actually use this

```suggestion
```

and you use `schema.example`.

File: src/components/DocumentView/OpenAPI/style.css
Comment (SamyPesse) [inline src/components/DocumentView/OpenAPI/style.css:176]: Why are we using a larger text font than the description?

It doesn't look good in the screenshot.

**Review by SamyPesse — CHANGES_REQUESTED** [review]
IMO it can't be merged in the current state and it requires more work:
- Properly support non-primitive values?
  - should we represent with `JSON.stringify`? If yes, it'll not work well for XML/etc API
  - should we skip non-primitives using `typeof schema.example === 'string' || typeof schema.example === 'number' || typeof schema.example === 'boolean'`?
- Should we handle `examples` as well by selecting the first example or the one matching the media-type?

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:17]: @SamyPesse it looks like you got it reversed. `example` is for schema fields, and `examples` is for parameters
Screen Shot 2024-05-21 at 2 56 09 PM

https://swagger.io/docs/specification/adding-examples/

Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx:17]: I also think we should keep this PR just for the singular `example` and not support `examples` for now. What do you think?

Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:17]: Yes, but it doesn't change the fact that this prop is actually not used:

```suggestion
```

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:96]: It's about the type. `schema.description` is always a string and if it's empty we don't want to show an empty div.

`schema.example` can be any value, depending on the schema.

File: packages/react-openapi/src/OpenAPISpec.tsx
Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISpec.tsx:51]: Why do we need this? Isn't `example` on the `parameter.schema` itself.

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:97]: Are you doing the changes here?

File: packages/react-openapi/src/OpenAPISpec.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISpec.tsx:51]: To maintain compatibility with Swagger 2.0 (in our docs we say we support 2.0). In 2.0, `example` is not on `parameter.schema` but on `parameter` itself.

In 3.0, `example` is on `parameter.schema`, so the destructuring happening in the line below will take of it

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx]: removed

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx]: done. this is what it looks like now
Screen Shot 2024-05-22 at 11 49 45 AM

File: src/components/DocumentView/OpenAPI/style.css
Comment (vibhanshub) [reply src/components/DocumentView/OpenAPI/style.css]: fixed. see the screenshot above

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx]: fixed

File: packages/react-openapi/src/OpenAPISpec.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISpec.tsx:51]: https://swagger.io/docs/specification/2-0/adding-examples/

**Comment by daskinw** [comment]
@SamyPesse this is blocking for an ICP customer going live with their API docs. I think Vib made the requested changes. Can we get it to move?

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISchema.tsx:99]: I mentioned it a few times already; but until we solve this part we should not merge this PR: this line is wrong.

`schema.example` can be any value; and the way you display it will not work with non-primitive values.

Example taken from the doc https://swagger.io/docs/specification/adding-examples/:

```yaml
paths:
  /users:
    post:
      summary: Adds a new user
      requestBody:
        content:
          application/json:
            schema:      # Request body contents
              type: object
              properties:
                id:
                  type: integer
                name:
                  type: string
              example:   # Sample object
                id: 10
                name: Jessica Smith
      responses:
        '200':
          description: OK
```

With this case, it'll display `Example: [object Object]` which is shit 💩 

This is a very common use case, so we need to handle it properly. I can help over a 30-minute pair programming session, or you can iterate on it.

**Review by SamyPesse — CHANGES_REQUESTED** [review]

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:99]: To explain more and why we can't merge it as is: you are solving it for one customer by handling only primitives; every customers with non-primitive examples (like in the one above) will be `Example: [object Object]` and will consider worst than not having it. Basically, this PR will be a net-negative for all users.

Comment (SamyPesse) [reply packages/react-openapi/src/OpenAPISchema.tsx:99]: Haa I see that you added the non-prmitive... ok let's merge but fix the boolean case first:

Otherwise it'll display `Example: `

Comment (SamyPesse) [inline packages/react-openapi/src/OpenAPISchema.tsx:99]:
```suggestion
                        <span className="openapi-schema-example">Example: <code>{JSON.stringify(schema.example)}</code></span>
```

**Review by SamyPesse — APPROVED** [review]

File: packages/react-openapi/src/OpenAPISchema.tsx
Comment (vibhanshub) [reply packages/react-openapi/src/OpenAPISchema.tsx]: Done. Will merge the PR now

**Pull request merged**
